Incompatibilities with NSS 3.81

Multiple tests have failed since NSS 3.81 was released in Fedora. [Downgrading to NSS 3.77](https://github.com/dogtagpki/jss/pull/883) seems to fix the problems.

https://github.com/dogtagpki/jss/runs/7506244588?check_suite_focus=true#step:7:2239
```
35/77 Test #35: SSLClientAuth .....................................***Failed    6.81 sec
Jul 25, 2022 6:27:06 PM org.mozilla.jss.CryptoManager initialize
INFO: JSS CryptoManager: successfully initialized from NSS database at /github/home/build/jss/results/nssdb
***FilePasswordCallback returns m1oZilla
using port:2876
Server about to create socket
Server created socket
Server specified cert directly
Server about to accept
client about to connect
Client specified cert directly
client connected
Server accepted
Server about to read
Error: Exception in thread "main" org.mozilla.jss.ssl.SSLSocketException: SSL_ForceHandshake failed: (-8010) Unknown error
	at org.mozilla.jss.ssl.SSLSocket.forceHandshake(Native Method)
	at org.mozilla.jss.tests.SSLClientAuth.testConnection(SSLClientAuth.java:372)
	at org.mozilla.jss.tests.SSLClientAuth.doIt(SSLClientAuth.java:183)
	at org.mozilla.jss.tests.SSLClientAuth.main(SSLClientAuth.java:117)
java.io.IOException: SocketException cannot read on socket: Error reading from socket: (-12271) SSL peer cannot verify your certificate.
	at org.mozilla.jss.ssl.SSLSocket.read(SSLSocket.java:1511)
	at org.mozilla.jss.ssl.SSLInputStream.read(SSLInputStream.java:43)
	at org.mozilla.jss.ssl.SSLInputStream.read(SSLInputStream.java:28)
	at org.mozilla.jss.tests.SSLClientAuth.run(SSLClientAuth.java:464)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.mozilla.jss.ssl.SSLSocketException: Error reading from socket: (-12271) SSL peer cannot verify your certificate.
	at org.mozilla.jss.ssl.SSLSocket.socketRead(Native Method)
	at org.mozilla.jss.ssl.SSLSocket.read(SSLSocket.java:1505)
	... 4 more
Server exiting
```
https://github.com/dogtagpki/jss/runs/7506244588?check_suite_focus=true#step:7:2305
```
53/77 Test #53: SSLEngine_RSA .....................................***Failed    7.18 sec
Initializing CryptoManager...
Jul 25, 2022 6:27:34 PM org.mozilla.jss.CryptoManager initialize
INFO: JSS CryptoManager: successfully initialized from NSS database at /github/home/build/jss/results/nssdb
Jul 25, 2022 6:27:35 PM org.mozilla.jss.util.Password finalize
WARNING: Uncleared Password object created at org.mozilla.jss.JSSLoader.parsePasswords(JSSLoader.java:327)
        called by org.mozilla.jss.JSSLoader.init(JSSLoader.java:125)
        called by org.mozilla.jss.JSSLoader.init(JSSLoader.java:89)
        called by org.mozilla.jss.JSSProvider.configure(JSSProvider.java:77)
        called by java.base/sun.security.jca.ProviderConfig$3.run(ProviderConfig.java:257)
        called by java.base/sun.security.jca.ProviderConfig$3.run(ProviderConfig.java:248)
        called by java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
        called by java.base/sun.security.jca.ProviderConfig.doLoadProvider(ProviderConfig.java:248)
        called by java.base/sun.security.jca.ProviderConfig.getProvider(ProviderConfig.java:226)
        called by java.base/sun.security.jca.ProviderList.getProvider(ProviderList.java:268)
        called by java.base/sun.security.jca.ProviderList.getIndex(ProviderList.java:298)
        called by java.base/sun.security.jca.ProviderList.getProviderConfig(ProviderList.java:282)
        called by java.base/sun.security.jca.ProviderList.getProvider(ProviderList.java:288)
        called by java.base/java.security.Security.getProvider(Security.java:474)
        called by org.mozilla.jss.CryptoManager.getInstance(CryptoManager.java:395)
        called by org.mozilla.jss.tests.TestSSLEngine.initialize(TestSSLEngine.java:45)
        called by org.mozilla.jss.tests.TestSSLEngine.main(TestSSLEngine.java:1014)
Please report the above trace to your software vendors.
...
Jul 25, 2022 6:27:41 PM org.mozilla.jss.ssl.javax.JSSEngineReferenceImpl info
INFO: [JSS Client TLSv1.2/TLS_DHE_RSA_WITH_AES_128_CBC_SHA256] JSSEngine.unwrap() - Got SSLException: javax.net.ssl.SSLException: Unexpected return from PR.Read(): SSL_ERROR_HANDSHAKE_FAILED (-12116)
Error: Exception in thread "main" javax.net.ssl.SSLException: Unexpected return from PR.Read(): SSL_ERROR_HANDSHAKE_FAILED (-12116)
	at org.mozilla.jss.ssl.javax.JSSEngineReferenceImpl.unwrap(JSSEngineReferenceImpl.java:1259)
	at java.base/javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:679)
	at org.mozilla.jss.tests.TestSSLEngine.testHandshake(TestSSLEngine.java:337)
	at org.mozilla.jss.tests.TestSSLEngine.testBasicHandshake(TestSSLEngine.java:779)
	at org.mozilla.jss.tests.TestSSLEngine.testAllHandshakes(TestSSLEngine.java:853)
	at org.mozilla.jss.tests.TestSSLEngine.testNativeClientServer(TestSSLEngine.java:1000)
	at org.mozilla.jss.tests.TestSSLEngine.main(TestSSLEngine.java:1029)
```

https://github.com/dogtagpki/jss/runs/7506244588?check_suite_focus=true#step:7:9650
```
60/77 Test #60: TestBufferPRFDSSL_RSA .............................***Failed    0.30 sec
error: SSL_ForceHandshake error -8010: SEC_ERROR_ALGORITHM_MISMATCH
Found cert with nickname: Server_RSA
Trying handshake...
Client Handshake:


Server Handshake:


Client Handshake:

      Start 61: TestBufferPRFDSSL_ECDSA
```

https://github.com/dogtagpki/jss/runs/7506244588?check_suite_focus=true#step:7:9665
```
62/77 Test #62: JSS_Test_BufferPRFD ...............................***Failed    0.68 sec
Calling TestCreateClose()...
Initializing CryptoManager...
Jul 25, 2022 6:27:45 PM org.mozilla.jss.CryptoManager initialize
INFO: JSS CryptoManager: successfully initialized from NSS database at /github/home/build/jss/results/nssdb
Calling TestSSLHandshake(Server_RSA, null)...
Initial: (TLS_1_2:32, TLS_1_3:35)
Actual: (TLS_1_2:32, TLS_1_3:35)
Initial: (TLS_1_2:32, TLS_1_3:35)
Actual: (TLS_1_2:32, TLS_1_3:35)
Unexpected error: SEC_ERROR_ALGORITHM_MISMATCH (-8010)
```

https://github.com/dogtagpki/jss/runs/7506244588?check_suite_focus=true#step:7:9686
```
67/77 Test #67: SSLClientAuth_FIPSMODE ............................***Failed    6.96 sec
Jul 25, 2022 6:27:53 PM org.mozilla.jss.CryptoManager initialize
INFO: JSS CryptoManager: successfully initialized from NSS database at /github/home/build/jss/results/fips
***FilePasswordCallback returns m1oZilla
using port:2877
The NSS database is configured in FIPSmode.
Enable ony FIPS ciphersuites.
Server about to create socket
Server created socket
Server specified cert directly
Server about to accept
client about to connect
Client specified cert directly
client connected
Server accepted
Server about to read
Error: Exception in thread "main" org.mozilla.jss.ssl.SSLSocketException: SSL_ForceHandshake failed: (-8010) Unknown error
	at org.mozilla.jss.ssl.SSLSocket.forceHandshake(Native Method)
	at org.mozilla.jss.tests.SSLClientAuth.testConnection(SSLClientAuth.java:372)
	at org.mozilla.jss.tests.SSLClientAuth.doIt(SSLClientAuth.java:183)
	at org.mozilla.jss.tests.SSLClientAuth.main(SSLClientAuth.java:117)
java.io.IOException: SocketException cannot read on socket: Error reading from socket: (-12271) SSL peer cannot verify your certificate.
	at org.mozilla.jss.ssl.SSLSocket.read(SSLSocket.java:1511)
	at org.mozilla.jss.ssl.SSLInputStream.read(SSLInputStream.java:43)
	at org.mozilla.jss.ssl.SSLInputStream.read(SSLInputStream.java:28)
	at org.mozilla.jss.tests.SSLClientAuth.run(SSLClientAuth.java:464)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.mozilla.jss.ssl.SSLSocketException: Error reading from socket: (-12271) SSL peer cannot verify your certificate.
	at org.mozilla.jss.ssl.SSLSocket.socketRead(Native Method)
	at org.mozilla.jss.ssl.SSLSocket.read(SSLSocket.java:1505)
	... 4 more
Server exiting
```

https://github.com/dogtagpki/jss/runs/7506244588?check_suite_focus=true#step:7:9728
````
72/77 Test #72: SSLEngine_RSA_FIPSMODE ............................***Failed    7.07 sec
Initializing CryptoManager...
Jul 25, 2022 6:28:08 PM org.mozilla.jss.CryptoManager initialize
INFO: JSS CryptoManager: successfully initialized from NSS database at /github/home/build/jss/results/fips
Jul 25, 2022 6:28:08 PM org.mozilla.jss.util.Password finalize
WARNING: Uncleared Password object created at org.mozilla.jss.JSSLoader.parsePasswords(JSSLoader.java:327)
        called by org.mozilla.jss.JSSLoader.init(JSSLoader.java:125)
        called by org.mozilla.jss.JSSLoader.init(JSSLoader.java:89)
        called by org.mozilla.jss.JSSProvider.configure(JSSProvider.java:77)
        called by java.base/sun.security.jca.ProviderConfig$3.run(ProviderConfig.java:257)
        called by java.base/sun.security.jca.ProviderConfig$3.run(ProviderConfig.java:248)
        called by java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
        called by java.base/sun.security.jca.ProviderConfig.doLoadProvider(ProviderConfig.java:248)
        called by java.base/sun.security.jca.ProviderConfig.getProvider(ProviderConfig.java:226)
        called by java.base/sun.security.jca.ProviderList.getProvider(ProviderList.java:268)
        called by java.base/sun.security.jca.ProviderList.getIndex(ProviderList.java:298)
        called by java.base/sun.security.jca.ProviderList.getProviderConfig(ProviderList.java:282)
        called by java.base/sun.security.jca.ProviderList.getProvider(ProviderList.java:288)
        called by java.base/java.security.Security.getProvider(Security.java:474)
        called by org.mozilla.jss.CryptoManager.getInstance(CryptoManager.java:395)
        called by org.mozilla.jss.tests.TestSSLEngine.initialize(TestSSLEngine.java:45)
        called by org.mozilla.jss.tests.TestSSLEngine.main(TestSSLEngine.java:1014)
Please report the above trace to your software vendors.
...
INFO: [JSS Client TLSv1.2/TLS_DHE_RSA_WITH_AES_128_CBC_SHA256] JSSEngine.unwrap() - Got SSLException: javax.net.ssl.SSLException: Unexpected return from PR.Read(): SSL_ERROR_HANDSHAKE_FAILED (-12116)
Error: Exception in thread "main" javax.net.ssl.SSLException: Unexpected return from PR.Read(): SSL_ERROR_HANDSHAKE_FAILED (-12116)
	at org.mozilla.jss.ssl.javax.JSSEngineReferenceImpl.unwrap(JSSEngineReferenceImpl.java:1259)
	at java.base/javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:679)
	at org.mozilla.jss.tests.TestSSLEngine.testHandshake(TestSSLEngine.java:337)
	at org.mozilla.jss.tests.TestSSLEngine.testBasicHandshake(TestSSLEngine.java:779)
	at org.mozilla.jss.tests.TestSSLEngine.testAllHandshakes(TestSSLEngine.java:853)
	at org.mozilla.jss.tests.TestSSLEngine.testNativeClientServer(TestSSLEngine.java:1000)
	at org.mozilla.jss.tests.TestSSLEngine.main(TestSSLEngine.java:1029)
```

--- a/base/src/test/java/org/mozilla/jss/tests/GenerateTestCert.java
+++ b/base/src/test/java/org/mozilla/jss/tests/GenerateTestCert.java
@@ -303,8 +303,13 @@ public class GenerateTestCert {
         int rand,
         SEQUENCE extensions) throws Exception {
 
-        AlgorithmIdentifier sigAlgID = new AlgorithmIdentifier(sigAlg.toOID());
-
+        AlgorithmIdentifier sigAlgID = null;
+        if(keyType.equals("RSA")) {
+            sigAlgID = new AlgorithmIdentifier(sigAlg.toOID(), null);
+        }
+        else {
+            sigAlgID = new AlgorithmIdentifier(sigAlg.toOID());
+        }
         Name issuer = new Name();
         issuer.addCountryName("US");
         issuer.addOrganizationName("Mozilla");

--- a/base/src/test/java/org/mozilla/jss/tests/SSLClientAuth.java
+++ b/base/src/test/java/org/mozilla/jss/tests/SSLClientAuth.java
@@ -73,7 +73,7 @@ public class SSLClientAuth implements Runnable {
     public static Certificate makeCert(String issuerName, String subjectName,
             int serialNumber, PrivateKey privKey, PublicKey pubKey, int rand,
             SEQUENCE extensions) throws Exception {
-        AlgorithmIdentifier sigAlgID = new AlgorithmIdentifier( sigAlg.toOID());
+        AlgorithmIdentifier sigAlgID = new AlgorithmIdentifier( sigAlg.toOID(), null);
 
         Name issuer = new Name();
         issuer.addCountryName("US");

